cmd/oras/blob/push.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package blob
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"oras.land/oras/cmd/oras/internal/display"
+	"oras.land/oras/cmd/oras/internal/option"
+	"oras.land/oras/internal/file"
+)
+
+type pushBlobOptions struct {
+	option.Common
+	option.Descriptor
+	option.Pretty
+	option.Remote
+
+	fileRef   string
+	targetRef string
+}
+
+func pushCmd() *cobra.Command {
+	var opts pushBlobOptions
+	cmd := &cobra.Command{
+		Use:   "push <name> file [flags]",
+		Short: "[Preview] Push a blob to a remote registry",
+		Long: `[Preview] Push a blob to a remote registry
+** This command is in preview and under development. **
+
+Example - Push blob "hi.txt":
+  oras blob push localhost:5000/hello hi.txt
+
+Example - Push blob from stdin:
+oras blob push localhost:5000/hello -
+
+Example - Push blob "hi.txt" and output the descriptor
+  oras blob push localhost:5000/hello hi.txt --descriptor
+
+Example - Push blob "hi.txt" and output the prettified descriptor
+  oras blob push localhost:5000/hello hi.txt --descriptor --pretty
+
+Example - Push blob without TLS:
+  oras blob push localhost:5000/hello hi.txt --insecure
+`,
+		Args: cobra.ExactArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			opts.fileRef = args[1]
+			if opts.fileRef == "-" && opts.PasswordFromStdin {
+				return errors.New("`-` read file from input and `--password-stdin` read password from input cannot be both used")
+			}
+			return opts.ReadPassword()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return pushBlob(opts)
+		},
+	}
+
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func pushBlob(opts pushBlobOptions) (err error) {
+	ctx, _ := opts.SetLoggerLevel()
+
+	repo, err := opts.NewRepository(opts.targetRef, opts.Common)
```

Comment:
`repo.Reference.Reference` should be empty or contains digest only. Otherwise, errors should be returned.

Think about
```
oras blob push localhost:5000/hello:v1 hi.txt
```
This does not look like a good syntax.

Reply:
This should be fine because the blob will be pushed to repository only. `repo.Reference.Reference` will be ignored when calling function `buildRepositoryBlobUploadURL`
(ref https://github.com/oras-project/oras-go/blob/main/registry/remote/repository.go#L613)
(ref https://github.com/oras-project/oras-go/blob/295b1c32b079ea18d9d49222a4a3d4d6d9c4b9ab/registry/remote/url.go#L86)

Reply:
Made requested change. Support push blob with specific digest.
```
Example - Push blob "hi.txt" with the specific digest:
  oras blob push localhost:5000/hello@sha256:9a201d228ebd966211f7d1131be19f152be428bd373a92071c71d8deaf83b3e5 hi.txt
```